### PR TITLE
interop: add keccak256 implementation

### LINKED ISSUE #3295: Implement Keccak256 functionality
Port https://github.com/neo-project/neo/pull/2925. Related to #3213.

### PR DESCRIPTION
**Pull request by AliceInHunterland**
Port neo-project/neo#2925.

Close #3295

### CODE REVIEW

**Comment by roman-khimov** [comment]
Should be a draft, we're not merging it before C# 3.7.

**Comment by AliceInHunterland** [comment]
problems with TestKeccak256_Compat

File: pkg/core/native/crypto.go
Comment (AnnaShaleva) [inline pkg/core/native/crypto.go:107]: It's not `sha256`.

File: pkg/core/native/crypto.go
Comment (AnnaShaleva) [inline pkg/core/native/crypto.go:302]: Wrap error into some meaningful.

File: pkg/core/native/crypto_test.go
Comment (AnnaShaleva) [inline pkg/core/native/crypto_test.go:32]: s/TestKeccak256/TestKeccak256_Compat
And add a corresponding comment with the link to the source of compat examples.

File: pkg/core/native/crypto_test.go
Comment (AnnaShaleva) [inline pkg/core/native/crypto_test.go:42]: Create the list of test-cases, each test-case is a structure, then in a single loop iterate through testcases. This allows to same some repetitive code. Take a look at TestStdLibItoaAtoi.

File: pkg/interop/native/crypto/crypto.go
Comment (AnnaShaleva) [inline pkg/interop/native/crypto/crypto.go:97]: Need a test for it on the compiler side.

**Review by AnnaShaleva — COMMENTED** [review]
Just for the record: once PR is finished and ready to be merged, we need to update interop deps and move it under Cockatrice hardfork. Depends on #3213.

File: pkg/core/native/crypto.go
Comment (AnnaShaleva) [reply pkg/core/native/crypto.go:107]: BTW, that's why you get problems with TestKeccak256_Compat.

File: pkg/interop/native/crypto/crypto.go
Comment (AliceInHunterland) [reply pkg/interop/native/crypto/crypto.go:97]: it's in pkg/compiler/native_test.go ?

Comment (AliceInHunterland) [reply pkg/interop/native/crypto/crypto.go:97]: i've added https://github.com/nspcc-dev/neo-go/pull/3301/commits/826162028a6523d920369209397a67ed126b1def#diff-38d3fe2c3c1c94577ac011836c3090e80ebe91c5fbc56aae38733a3b09a86626R242 but it has undefined: crypto.Keccak256 probably because of keccak256 not in the master

File: pkg/core/native/crypto.go
Comment (AnnaShaleva) [inline pkg/core/native/crypto.go:296]: Just `panic(err)` to match the rest of conversion code of other methods?

File: pkg/interop/native/crypto/crypto.go
Comment (AnnaShaleva) [reply pkg/interop/native/crypto/crypto.go:97]: Yes, add this test and after that create a separate commit with interop dependency update. Use `./scripts/update-deps.sh` to update the dependencies. See the example of such update: https://github.com/nspcc-dev/neo-go/pull/3155/commits/c6d259b28577725a971987157c0e0be37dd763da

After update test should pass.

File: pkg/core/native/crypto_test.go
Comment (AnnaShaleva) [inline pkg/core/native/crypto_test.go:40]: s/`expectPanic`/`fail`?

File: pkg/core/native/crypto_test.go
Comment (AnnaShaleva) [inline pkg/core/native/crypto_test.go:56]: But why interop?

Comment (AnnaShaleva) [reply pkg/core/native/crypto_test.go:56]: If this test structure doesn't suit for failing test cases handling, then a separate test for failing test cases can be added below.

Comment (AnnaShaleva) [reply pkg/core/native/crypto_test.go:56]: See TestStdLibItoaAtoi as an example.

File: pkg/core/native/crypto_test.go
Comment (AnnaShaleva) [reply pkg/core/native/crypto_test.go]: Link to C# examples is needed.

File: pkg/core/native/crypto_test.go
Comment (AnnaShaleva) [inline pkg/core/native/crypto_test.go:44]: Remove "bad arg type".

File: pkg/core/native/crypto_test.go
Comment (AnnaShaleva) [inline pkg/core/native/crypto_test.go:42]: Remove `fail`.

File: pkg/core/native/crypto_test.go
Comment (AnnaShaleva) [inline pkg/core/native/crypto_test.go:54]: Add `t.Run("good", ...)`.

File: pkg/core/native/crypto_test.go
Comment (AnnaShaleva) [inline pkg/core/native/crypto_test.go:67]: s/`keccaak errors`/`errors`, it's a single function, so no other options except `keccaak` are tested.

File: pkg/core/native/crypto_test.go
Comment (AnnaShaleva) [inline pkg/core/native/crypto_test.go:73]: It's `Null item` or just a `Null`.

File: pkg/core/native/crypto_test.go
Comment (AnnaShaleva) [inline pkg/core/native/crypto_test.go:74]: And this is an Interop. If you want a pure Null, then use `stackitem.Null{}`.

File: pkg/core/native/crypto_test.go
Comment (AnnaShaleva) [inline pkg/core/native/crypto_test.go:77]: `not a byte array`?

File: pkg/core/native/native_test/cryptolib_test.go
Comment (AnnaShaleva) [inline pkg/core/native/native_test/cryptolib_test.go:121]: s/inputData/in?

File: pkg/core/native/native_test/cryptolib_test.go
Comment (AnnaShaleva) [inline pkg/core/native/native_test/cryptolib_test.go:123]: Usually we have `expected` and `actual`.

File: pkg/core/native/native_test/cryptolib_test.go
Comment (AnnaShaleva) [inline pkg/core/native/native_test/cryptolib_test.go:134]: `actual := hex.EncodeToString(itm.Value().([]byte))`

You don't perform any other actions on these vars anyway.

**Review by AnnaShaleva — COMMENTED** [review]
Need to update interop deps one more time after PR is about to be merged to master.

File: pkg/interop/native/crypto/crypto.go
Comment (AnnaShaleva) [inline pkg/interop/native/crypto/crypto.go:96]: Preserve the line width.

**Review by AnnaShaleva — COMMENTED** [review]
Otherwise LGTM, will keep this until some real proximity to v3.7.0.

**Comment by AnnaShaleva** [comment]
@AliceInHunterland, please, rebase this PR onto fresh master and update go.mod files in a separate commit. We'll merge this PR after update, 3.7 is coming.

**Comment by AnnaShaleva** [comment]
@AliceInHunterland, tests are failing. Please, convert from draft when ready.

**Comment by AnnaShaleva** [comment]
@AliceInHunterland, ditto, block 20 stateroot from RPC tests needs to be updated because you change native Management's state by adding new method to native CryptoLib's manifest.

**Review by AnnaShaleva — APPROVED** [review]

**Pull request merged**
